### PR TITLE
[Bugfix][Attention] Enable FlashInfer mm-prefix attention

### PR DESCRIPTION
**Pull request by lesj0610**
## Purpose

Allow eligible multimodal PrefixLM / mm-prefix batches to stay on the native FlashInfer attention backend. This PR is the vLLM adapter for FlashInfer's mm-prefix batch-prefill wrapper: it builds the per-query ranges from request metadata, decides when the path is eligible, and calls the wrapper. The attention variant, its JIT generator, and the kernel-level tests live in FlashInfer; range construction, ubatch re-keying, eligibility and fallback, and wrapper invocation live here.

No mask is materialized. A dense boolean mask over the attention rectangle costs `O(qo_len * kv_len)` per step, so 16384 scheduled tokens against a 262144-token context is a 4 GiB mask. Nothing here builds one. vLLM passes a single `int32 (N, 2)` buffer of compact per-query `[start, end]` rows, sized by scheduled tokens and never by context length, and the wrapper's attention variant evaluates the whole expression in-kernel per KV tile:

```text
(causal AND sliding_window) OR mm_prefix
```

Those ranges are 128 KiB at 16384 scheduled tokens and do not grow with context.

Dependency: FlashInfer upstream PR 5189, now merged. The wrapper is in FlashInfer main, so it arrives with the next nightly and reaches a release once vLLM bumps its FlashInfer pin. The pinned release does not have it yet, so `supports_mm_prefix()` returns `False` and the backend is demoted before selection; support is advertised only after the module builds successfully for the requested configuration.

On SM100 the composite multimodal-prefix backend that landed in main is offered ahead of FlashInfer whenever a batch carries mm-prefix ranges. This PR puts FlashInfer first there: it serves the whole batch in one kernel when its own combination validates, and the composite stays the fallback for what it rejects. Selection validates all candidates and chooses the valid backend with the highest priority, so configurations FlashInfer rejects still fall through to the composite.

## Changes

- Construct FlashInfer's mm-prefix batch-prefill wrapper, resolved by name at call time rather than imported at module scope. A FlashInfer without it then demotes this backend before selection, where a module-scope import would raise `ImportError` for every user of the backend. `supports_mm_prefix()` is advertised only after the capability probe has built the module once through FlashInfer's public generator, and `supports_combination()` re-probes per layer group with that group's own head size and resolved KV dtype, since hybrid models such as Gemma4 select their backend per group. `FLASHINFER_DISABLE_JIT`, a missing toolchain, or a FlashInfer without the wrapper therefore demote the backend rather than crash startup after it, and the builder re-checks its exact combination at engine start as a fail-fast backstop.
- Build the per-query range metadata only when `fill_mm_prefix_query_ranges()` reports covered tokens (`> 0`), so a text-only or decode-only batch keeps the plain causal path. Hand it over as the `(N, 2)` rows the wrapper's public contract requires, slicing the prefill portion as `mm_ranges_gpu[num_decode_tokens:]`: the wrapper rejects a non-contiguous range tensor rather than copying it, and a row slice of this row-major buffer is contiguous. Pass `q_scale`, `k_scale` and `v_scale` to `run()` instead of folding `q_scale * k_scale` into an `sm_scale` scalar by hand. Keep the plain causal wrapper planned alongside the mm wrapper so forward time dispatches between them on `mm_prefix_query_range_tensor`, honoring Gemma4's late clear of that field for its full-attention layer groups.
- Classify short chunked-prefill extends as prefills on mm batches (`treat_short_extends_as_decodes=False`) so a tail inside a bidirectional range keeps access to the future tokens of its range. `mm_req_doc_ranges` and `is_prefilling` are already on `CommonAttentionMetadata`, but the ubatch split did not carry them, so the range dict is re-keyed to ubatch-local request indices and `is_prefilling` is sliced per side.
- Serve unquantized caches (`auto`/`float16`/`bfloat16`) and reject FP8, `nvfp4_4over6`, and `nvfp4` on SM100 for the mm-prefix path; this PR does not claim to enable an end-to-end selectable mm-prefix NVFP4 configuration. The remaining unsupported combinations fail closed: attention sinks, DCP, `logits_soft_cap > 0`, explicitly forced trtllm-gen attention, and trtllm-only kernel page sizes (`>= 128` rejected at selection time, so large pages are no longer advertised for mm-prefix models).
- Offer `FLASHINFER` before `TRITON_FLASHINFER` in the SM100 priority list when `use_mm_prefix` is set. The composite reaches these batches by routing the bidirectional spans to Triton and the causal remainder to FlashInfer, while FlashInfer does both in one kernel when it validates. The composite still answers for an fp8 cache, a trtllm-only page size, or a build whose module will not compile. Forward `use_rswa` through `FlashInferBackend.validate_configuration()`: this is a compatibility fix, not R-SWA support, because the base method gained the argument when the composite landed and this backend overrides it, so without the forward every selection through FlashInfer raised `TypeError`.

## Test Plan

The adapter tests need a FlashInfer that provides the wrapper, so point at one built from FlashInfer main. Against the pinned release the wrapper-dependent tests skip instead.

```bash
FLASHINFER_PR_CHECKOUT=/path/to/flashinfer-checkout

CUDA_VISIBLE_DEVICES=0 \
PYTHONPATH="$FLASHINFER_PR_CHECKOUT" \
pytest \
  tests/v1/attention/test_mm_prefix.py \
  tests/v1/attention/test_attention_splitting.py \
  -q -rs
```

The selector and override checks do not need it:

```bash
pytest \
  tests/kernels/attention/test_attention_selector.py::test_hopper_mm_prefix_selects_triton_flash_attn \
  tests/kernels/attention/test_attention_selector.py::test_mm_prefix_priority_without_changing_causal_default \
  tests/v1/attention/test_mm_prefix.py::test_flashinfer_mm_prefix_validate_configuration_rejects_dcp \
  -q -rs
```

The two selector tests that enumerate the mm-prefix fallback order take FlashInfer eligibility as a parameter and assert the order either way: FlashInfer when its combination validates, the composite or the Triton fallback when it does not.

What the FlashInfer path is covered for:

- End-to-end builder → impl → wrapper against a dense float32 reference on paged KV, with a sliding window narrower than the widest range so the clamp-off and clamp-on formulas genuinely differ and both are checked.
- The `(N, 2)` shape and contiguity of the prefill range slice, and the cleared-metadata fallback to the causal wrapper.
- Capability rejections (NVFP4, FP8, sinks, DCP, page size, wrapper unavailable), per-group probe arguments for hybrid head sizes, kernel block-size advertising for mm and non-mm models, builder fail-fast on soft-cap and forced trtllm, and text-only and decode-only gating.
- Short-extend reclassification and DBO slicing, covering field preservation and a split request whose per-side fills must reproduce the unsplit assignment, inclusive end included.

The kernel-level tests — the mask against a dense reference, window semantics, partial query tiles, packed NVFP4 — moved to FlashInfer with the variant.

## Test Result

On one SM80 device, with the FlashInfer branch that carries the wrapper:

```text
pytest tests/v1/attention/test_mm_prefix.py tests/v1/attention/test_attention_splitting.py -q -rs
45 passed, 22 skipped
```

All 22 skips are hardware: 10 pre-existing FA4-only cases, and 12 composite cases that need Blackwell or Hopper. No FlashInfer adapter test skips. The targeted selector and override checks completed with `21 passed`, covering native FlashInfer availability in both directions, the Hopper and SM100 priority order, and `use_rswa` forwarding through the FlashInfer override.

On a FlashInfer without the wrapper the capability probe returns `False` and `supports_combination()` returns the rejection that backend selection consumes, so this backend is demoted before it is chosen. That path is covered by the selector tests rather than by a serving run; no fallback serving was executed for this revision.

`ruff check` / `ruff format --check` passed; commit hooks passed, including `mypy` for Python 3.10.

<details>
<summary>Historical serving measurements from the earlier in-tree variant</summary>

These results are historical context, not validation of the current wrapper-based adapter. They were measured with the former in-tree variant on a tree that also carried persistent attention-workspace profiling. That profiling code is not part of this PR, and the measurements have not been repeated against this revision.

Run on 2 SM80 devices, `-tp 2`, BF16 weights and KV, a 31B multimodal model, `--max-model-len 8192`, `--max-num-batched-tokens 8192`, `--max-num-seqs 128`, `--gpu-memory-utilization 0.85`. FlashInfer carried the mm-prefix batches end to end; a single 896x896 image request completed successfully, and no workspace or KV-cache assertion fired. The same-host comparison against the `TRITON_ATTN` baseline used the image workload definition and seeds from the composite PR's repro block: an 896x896 image plus a configured 1024-token text input and 1024-token output, with multimodal expansion producing 1295-1297 measured input tokens, `--temperature 0 --ignore-eos`, and prefix and mm caches reset between runs. The server configurations differ in attention backend, block size, and available KV capacity, so this compares complete serving configurations rather than isolating the mm-prefix kernel. **This is not a reproduction of that PR's numbers** — different architecture and parallelism, so only the within-host results are meaningful.

Values are `TRITON_ATTN / FLASHINFER`, **bold is better** (higher tok/s, lower latency):

| Conc | Total tok/s | Output tok/s | TTFT (ms) | TPOT (ms) | E2EL (ms) |
|---|---|---|---|---|---|
| 1 | 14.2 / **58.5** | 6.3 / **25.8** | 3884 / **2167** | 155.4 / **36.6** | 162860 / **39652** |
| 32 | **437.8** / 426.7 | **193.3** / 188.4 | 31441 / **30559** | **134.9** / 140.1 | **169446** / 173846 |
| 128 | 416.2 / **438.0** | 183.7 / **193.4** | 280114 / **275765** | **186.1** / 190.2 | 470526 / **470314** |

Caveats, so the table is not read for more than it says:

- The concurrency-32 FlashInfer figure is the mean of three seeds (10032/10033/10034: 426.6 / 427.0 / 426.7, spread 0.09%). The `TRITON_ATTN` figure on that row is a **single seed**, so its 2.6% lead carries an unmeasured seed-variance error bar and should not be read as a real difference. Concurrency 1 and 128 are single-seed on both sides.
- At concurrency 32, the measured inputs are 1295-1297 tokens with 1024 output tokens, putting the maximum full-request footprint at approximately 74K tokens, above either KV budget; concurrency 128 is more constrained. Scheduling and cache pressure can therefore affect these rows, and they do not isolate attention-kernel cost.
- The two arms do not get identical KV budgets: FlashInfer reserves its attention workspace during profiling and lands at 49,593 tokens against the baseline's 51,515. The FlashInfer configuration is therefore measured with approximately 3.7% less KV capacity.

</details>

## Notes

- When the native FlashInfer mm-prefix backend is eligible, its execution path uses native fa2 prefill; trtllm-gen and cascade attention are not selected for those batches.
- mm-prefix batches pay full-rectangle KV traversal on the prefill, the same compute shape the dense custom mask had. The win of this rework is memory, not FLOPs: no `O(qo_len * kv_len)` mask is ever materialized.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results.
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. Not needed: no user-facing model or config surface changes.
</details>

AI assistance: Codex, Claude

### CODE REVIEW

**Comment by geunamja-gimhae** [comment]
**Real-world impact report + offer to test on SM120 (production Gemma-4 multimodal)**

We run vLLM in production serving `gemma-4-26B-A4B-it-FP8` (multimodal, `is_mm_prefix_lm=True`) on 2× RTX 5090 (SM120), TP=2, fp8 KV cache, `ngram_gpu` speculative decoding (k=8), mostly long-context RAG traffic.

Because the FlashInfer backend does not declare `supports_mm_prefix`, backend selection disqualifies it for this model and falls back to `TRITON_ATTN` — for every request, including the 100%-text traffic that dominates our workload. Verified on v0.25.1 and again on v0.28.0 (boot log: `Using AttentionBackendEnum.TRITON_ATTN backend.`). `--limit-mm-per-prompt '{"image": 0, "video": 0}'` does not lift the gate, since `is_mm_prefix_lm` is a model-architecture property rather than an input question (and attempting that flag with a warm compile cache also runs into the cache-key collision from #50891). So there is currently no configuration that lets an mm-prefix model reach the SM12x XQA decode path added in #49718.

The cost on consumer Blackwell is large: flashinfer-ai/flashinfer#4137 measured XQA against vLLM's Triton unified-attention kernel on RTX PRO 6000 (SM120) at 2.4–3.6× for m=1 decode, and up to 14–18× for spec-decode verification (m=4/8) on full-attention layers at 32k context — exactly our decode profile. Today, mm-prefix models are excluded from all of it.

This PR would unblock a real production speedup for Gemma-4-class multimodal models. If it helps, we are happy to run correctness and throughput A/B on our 2× RTX 5090 setup once the PR is in a reviewable state — we have 0.25.1 baselines and a test harness ready.

*Disclosure: drafted with AI assistance; the deployment details and observations above are from our own production system.*

**Comment by lesj0610** [comment]
@DarkLight1337 Sorry for the cold ping — I've run out of ideas on who to ask.

This PR has been open since June 24. I reworked the implementation in August, and in late August another user reported being hit by the same issue and suggested a test case, so it isn't stale on my end — it's just never had a maintainer look at it.

I'll be honest: keeping this branch current has become hard to sustain. A good part of my time on it now goes to conflict resolution rather than the change itself.

I'm also looking for direction on #46883 and #46963, which sit in the same area. Some guidance on whether these approaches are worth pursuing would help me decide where to put my time. For this one, feedback on the current approach — or a pointer to the right reviewer — would mean a lot.

**Comment by DarkLight1337** [comment]
Please ask @mgoin for review

**Comment by lesj0610** [comment]
> Please ask @mgoin for review

Thanks. @mgoin — over to you then.

Open since June 24. It removes the forced Triton fallback for the mm-prefix path; I reworked the implementation in August, and another user ran into the same issue in late August and suggested a test case.

#46883 and #46963 sit in the same area if it's easier to take them together.

**Comment by lesj0610** [comment]
@pavanimajety @vadiklyutiy — you both own `vllm/v1/attention/backends/flashinfer.py` in CODEOWNERS and this PR has not been reviewed yet, so bringing it to you directly.

What changed since the earlier revision:

- The FlashInfer dependency it needs (`variant_owns_mask`) is merged upstream, so the JIT variant no longer depends on an unlanded change.
- The description's "not run" gap is closed with an end-to-end multimodal serving run on SM80 with `-tp 2`, plus a same-host `TRITON_ATTN` comparison. The table explicitly documents the seed asymmetry, server-configuration differences, and unequal KV budgets; it is presented as an end-to-end configuration comparison rather than an isolated kernel measurement.
- Merge conflicts are resolved; the branch is mergeable.

The one thing I would value your read on most: `supports_mm_prefix()` advertises the backend only after the fa2 JIT variant has actually been built once, and re-probes per layer group in `supports_combination()` with that group's own head size and resolved KV dtype. That is the part that decides whether a hybrid model demotes the backend cleanly or crashes after selection, and it is where a second pair of eyes would help most.

Happy to split this into smaller pieces if the size is the obstacle.

File: vllm/v1/attention/backends/flashinfer.py [hunk 1]
```diff
@@ -99,6 +103,154 @@
...
+    const int32_t s = params.mm_q_ranges[row * 2];
+    const int32_t e = params.mm_q_ranges[row * 2 + 1];
+    bool mm = (kv >= s) && (kv <= e);
+    if (clamp_sw > 0) {
+      mm = mm && ((q_abs - kv) < clamp_sw);
+    }
+    return keep || mm;
+  })
+};
+"""
```
Comment (Isotr0py) [inline vllm/v1/attention/backends/flashinfer.py]: This is too hacky, you should submit a PR at FI side to integrate it firstly instead.

**Review by Isotr0py — CHANGES_REQUESTED** [review]

File: vllm/v1/attention/backends/flashinfer.py (hunk 1 above)
Comment (lesj0610) [reply vllm/v1/attention/backends/flashinfer.py]: You're right — this belongs in FlashInfer. I've opened it as FlashInfer upstream PR 5189.

That PR contains the attention variant, JIT generator, public wrapper, and kernel-level tests. I'm reworking this PR against that API now: what remains here is constructing and re-keying the ranges, backend selection and fallback, and the wrapper call.

Comment (lesj0610) [reply vllm/v1/attention/backends/flashinfer.py]: I've pushed the smaller version. The variant and JIT ABI now live in FlashInfer upstream PR 5189. This PR is down to the vLLM side: building and re-keying the ranges, checking backend eligibility, and calling the wrapper.

The composite backend landed on main in the meantime. On SM100 this branch now tries native FlashInfer first and keeps the composite immediately behind it for combinations FlashInfer rejects.

**Comment by lesj0610** [comment]
@Isotr0py The FlashInfer side landed. Upstream PR 5189 is merged, so the attention variant, its JIT generator and the kernel-level tests live there now instead of in this PR.

What is left here is the vLLM adapter: building the ranges from request metadata, deciding when the path is eligible, and calling the wrapper. It is a no-op against the pinned FlashInfer release, which does not carry the wrapper yet, so `supports_mm_prefix()` returns `False` and the backend is demoted before selection. It becomes live once the pin moves.
